packages/astro/src/content/vite-plugin-content-imports.ts

```diff
@@ -41,116 +45,201 @@ export function astroContentImportPlugin({
-
-				const code = escapeViteEnvReferences(`
-export const id = ${JSON.stringify(generatedInfo.id)};
-export const collection = ${JSON.stringify(generatedInfo.collection)};
+					const code = escapeViteEnvReferences(`
+export const id = ${JSON.stringify(id)};
+export const collection = ${JSON.stringify(collection)};
 export const slug = ${JSON.stringify(slug)};
-export const body = ${JSON.stringify(info.body)};
+export const body = ${JSON.stringify(body)};
 export const data = ${devalue.uneval(data) /* TODO: reuse astro props serializer */};
 export const _internal = {
 	filePath: ${JSON.stringify(_internal.filePath)},
 	rawData: ${JSON.stringify(_internal.rawData)},
 };
 `);
-				return { code };
-			}
-		},
-		configureServer(viteServer) {
-			viteServer.watcher.on('all', async (event, entry) => {
-				if (
-					['add', 'unlink', 'change'].includes(event) &&
-					getEntryType(entry, contentPaths, contentEntryExts) === 'config'
-				) {
-					// Content modules depend on config, so we need to invalidate them.
-					for (const modUrl of viteServer.moduleGraph.urlToModuleMap.keys()) {
-						if (isContentFlagImport(modUrl, contentEntryExts)) {
-							const mod = await viteServer.moduleGraph.getModuleByUrl(modUrl);
-							if (mod) {
-								viteServer.moduleGraph.invalidateModule(mod);
+					return { code };
+				}
+			},
+			configureServer(viteServer) {
+				viteServer.watcher.on('all', async (event, entry) => {
+					if (
+						['add', 'unlink', 'change'].includes(event) &&
+						getEntryType(entry, contentPaths, contentEntryExts) === 'config'
+					) {
+						// Content modules depend on config, so we need to invalidate them.
+						for (const modUrl of viteServer.moduleGraph.urlToModuleMap.keys()) {
+							if (
+								isContentFlagImport(modUrl, contentEntryExts) ||
+								// TODO: refine to content types with getModule
+								contentEntryExts.some((ext) => modUrl.endsWith(ext))
+							) {
+								const mod = await viteServer.moduleGraph.getModuleByUrl(modUrl);
+								if (mod) {
+									viteServer.moduleGraph.invalidateModule(mod);
+								}
 							}
 						}
 					}
+				});
+			},
+			async transform(code, id) {
+				if (isContentFlagImport(id, contentEntryExts)) {
+					// Escape before Rollup internal transform.
+					// Base on MUCH trial-and-error, inspired by MDX integration 2-step transform.
+					return { code: escapeViteEnvReferences(code) };
 				}
-			});
+			},
 		},
-		async transform(code, id) {
-			if (isContentFlagImport(id, contentEntryExts)) {
-				// Escape before Rollup internal transform.
-				// Base on MUCH trial-and-error, inspired by MDX integration 2-step transform.
-				return { code: escapeViteEnvReferences(code) };
+	];
+
+	if (settings.contentEntryTypes.some((t) => t.getRenderModule)) {
+		plugins.push({
+			name: 'astro:content-render-imports',
+			async load(viteId) {
+				if (!contentEntryExts.some((ext) => viteId.endsWith(ext))) return;
```

Comment:
Deserves a comment! We're ensuring this module should be handled by content collections at all. If not, short circuit to avoid processing the `fileId`.